File: CSharp/ExtensibleCommands/ExtensibleCommandsUnitTests/ParallelCommandTest.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using Microsoft.VisualStudio.TestTools.UnitTesting;

using System.Threading;
using System.Linq;

using ExtensibleCommands;

namespace ExtensibleCommandsUnitTest
{
    /// <summary>
    ///This is a test class for ParallelCommandTest and is intended
    ///to contain all ParallelCommandTest Unit Tests
    ///</summary>
    [TestClass()]
    public class ParallelCommandTest
    {
        private TestContext testContextInstance;

        /// <summary>
        ///Gets or sets the test context which provides
        ///information about and functionality for the current test run.
        ///</summary>
        public TestContext TestContext
        {
            get
            {
                return testContextInstance;
            }
            set
            {
                testContextInstance = value;
            }
        }

        #region Additional test attributes
        // 
        //You can use the following additional attributes as you write your tests:
        //
        //Use ClassInitialize to run code before running the first test in the class
        [ClassInitialize()]
        public static void MyClassInitialize(TestContext testContext)
        {
            Setup.InitLog();
        }
        //
        //Use ClassCleanup to run code after all tests in a class have run
        //[ClassCleanup()]
        //public static void MyClassCleanup()
        //{
        //}
        //
        //Use TestInitialize to run code before running each test
        [TestInitialize()]
        public void MyTestInitialize()
        {
            Logger.Log(Logger.LogLevel.Info,
                "----------------------------------------------------------------------------------------------------------");
            Logger.Log(Logger.LogLevel.Info,
                string.Format("Starting Test {0}:{1}", GetType().Name, testContextInstance.TestName));
        }
        //
        //Use TestCleanup to run code after each test has run
        //[TestCleanup()]
        //public void MyTestCleanup()
        //{
        //}
        //
        #endregion

        [TestMethod()]
        public void ModifyWhileExecutingTest()
        {
            bool exceptionCaught = false;
            var command = new ParallelCommand("P")
                .Add(new AbortableSleepCommand(Setup.ThreadLatencyDelayMsec))
                .Add(new AbortableSleepCommand(Setup.ThreadLatencyDelayMsec));

            new Thread(command.Run).Start();
            Thread.Sleep((int)(0.5 * Setup.ThreadLatencyDelayMsec));

            try
            {
                command.Add(SimpleCommand.NullCommand);
            }
            catch (Exception e)
            {
                if (e.Message.Contains("Attempt to add"))
                    exceptionCaught = true;
            }
            Assert.IsTrue(exceptionCaught);
        }

        [TestMethod()]
        public void RunOKTest()
        {
            var command = new ParallelCommand("Parallel")
                .Add(new SimpleCommand(() => Thread.Sleep(200), "P1"))
                .Add(new SimpleCommand(() => Thread.Sleep(300), "P2"))
                .Add(new SimpleCommand(() => Thread.Sleep(400), "P3"))
                .Add(new SimpleCommand(() => Thread.Sleep(500), "P4"))
                .Add(new SimpleCommand(() => Thread.Sleep(600), "P5"))
                .Add(new SimpleCommand(() => Thread.Sleep(600), "P6"))
                .Add(new SimpleCommand(() => Thread.Sleep(600), "P7"))
                .Add(new SimpleCommand(() => Thread.Sleep(600), "P8"))
                .Add(new SimpleCommand(() => Thread.Sleep(600), "P9"));

            Setup.RunAndWaitForNormalCompletion(command);

            Assert.IsTrue(command.ElapsedTimeMsec < 1500);
        }

        [TestMethod()]
        public void RunErrorTest()
        {
            var command = new ParallelCommand("Parallel")
                .Add(new SimpleCommand(() => Thread.Sleep(200), "P1"))
                .Add(new SimpleCommand(() => Thread.Sleep(300), "P2"))
                .Add(new SimpleCommand(() => { throw new ExtensibleCommandsException(Setup.TestErrorCode, Setup.TestErrorDescription); }, "P3-Error"))
                .Add(new SimpleCommand(() => Thread.Sleep(500), "P4"))
                .Add(new SimpleCommand(() => Thread.Sleep(600), "P5"))
                .Add(new SimpleCommand(() => Thread.Sleep(600), "P6"))
                .Add(new SimpleCommand(() => Thread.Sleep(600), "P7"));

            Setup.RunAndWaitForFailure(command);

            Assert.IsTrue(command.ElapsedTimeMsec < 1500);
            Assert.AreEqual(State.Completed, command.GetSubCommand(0).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(1).CurrentState);
            Assert.AreEqual(State.Failed, command.GetSubCommand(2).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(3).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(4).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(5).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(6).CurrentState);

            Assert.AreEqual(Setup.TestErrorCode, command.Exception.ID);
            Assert.AreEqual(Setup.TestErrorDescription, command.Exception.Text);
        }

        [TestMethod()]
        public void RunFatalErrorTest()
        {
            var command = new ParallelCommand("Parallel")
                .Add(new SimpleCommand(() => Thread.Sleep(200), "P1"))
                .Add(new SimpleCommand(() => Thread.Sleep(300), "P2"))
                .Add(new SimpleCommand(() => { throw new Exception(Setup.TestErrorDescription); }, "P3-Error"))
                .Add(new SimpleCommand(() => Thread.Sleep(500), "P4"))
                .Add(new SimpleCommand(() => Thread.Sleep(600), "P5"))
                .Add(new SimpleCommand(() => Thread.Sleep(600), "P6"))
                .Add(new SimpleCommand(() => Thread.Sleep(600), "P7"));

            try
            {
                command.Run();
            }
            catch (Exception ex)
            {
                Assert.IsTrue(ex.Message.StartsWith("Fatal error"));
                Assert.AreEqual(ex.InnerException.Message, Setup.TestErrorDescription);
            }

            Assert.IsTrue(command.ElapsedTimeMsec < 1500);
            Assert.AreEqual(State.Completed, command.GetSubCommand(0).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(1).CurrentState);
            Assert.AreEqual(State.Failed, command.GetSubCommand(2).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(3).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(4).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(5).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(6).CurrentState);
            Assert.IsNull(command.Exception);
        }

        [TestMethod()]
        public void AbortTest()
        {
            var command = new ParallelCommand();
            command.Add(new SimpleCommand(() => Thread.Sleep(100), "P1"))
                .Add(new SimpleCommand(() => Thread.Sleep(100), "P2"))
                .Add(new SimpleCommand(() => command.Abort(), "P3-Abort"))
                .Add(new SimpleCommand(() => Thread.Sleep(500), "P4"))
                .Add(new SimpleCommand(() => Thread.Sleep(400), "P5"));

            Setup.RunAndWaitForAbort(command);

            Assert.AreEqual(State.Aborted, command.CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(0).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(1).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(2).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(3).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(4).CurrentState);
        }

        [TestMethod()]
        public void AbortTestWithAbortableSubCommands()
        {
            var command = new ParallelCommand();
            command.Add(new AbortableSleepCommand(2600, "P1"))
                .Add(new AbortableSleepCommand(2700, "P2"))
                .Add(new SimpleCommand(() => { Thread.Sleep(3*Setup.ThreadLatencyDelayMsec); command.Abort(); }, "P3-Abort"))
                .Add(new AbortableSleepCommand(3700, "P4"));

            Setup.RunAndWaitForAbort(command);
            Assert.AreEqual(State.Aborted, command.CurrentState);
            Assert.AreEqual(State.Aborted, command.GetSubCommand(0).CurrentState);
            Assert.AreEqual(State.Aborted, command.GetSubCommand(1).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(2).CurrentState);
            Assert.AreEqual(State.Aborted, command.GetSubCommand(3).CurrentState);
        }

        [TestMethod()]
        public void PauseResumeTest()
        {
            var command = CreatePauseAbortParallelCommand();

            var assertAfterPause = new Action(() =>
            {
                Assert.IsTrue(command.CurrentState == State.Executing);
                Assert.AreEqual(State.Completed, command.GetSubCommand(0).CurrentState);
                Assert.AreEqual(State.Completed, command.GetSubCommand(1).CurrentState);
                Assert.AreEqual(State.Executing, command.GetSubCommand(2).CurrentState);
                Assert.AreEqual(State.Executing, command.GetSubCommand(3).CurrentState);
                Assert.AreEqual(State.Executing, command.GetSubCommand(4).CurrentState);
                Assert.AreEqual(State.Executing, command.GetSubCommand(5).CurrentState);
            });

            Setup.PauseAndResume(command, assertAfterPause);

            Assert.AreEqual(State.Completed, command.CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(0).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(1).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(2).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(3).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(4).CurrentState);
            Assert.AreEqual(State.Completed, command.GetSubCommand(5).CurrentState);
        }

        [TestMethod()]
        public void PauseAbortTest()
        {
            var command = CreatePauseAbortParallelCommand();

            var assertAfterPause = new Action(() =>
            {
                Assert.IsTrue(command.CurrentState == State.Executing);
                Assert.AreEqual(State.Completed, command.GetSubCommand(0).CurrentState);
                Assert.AreEqual(State.Completed, command.GetSubCommand(1).CurrentState);
                Assert.AreEqual(State.Executing, command.GetSubCommand(2).CurrentState);
                Assert.AreEqual(State.Executing, command.GetSubCommand(3).CurrentState);
                Assert.AreEqual(State.Executing, command.GetSubCommand(4).CurrentState);
                Assert.AreEqual(State.Executing, command.GetSubCommand(5).CurrentState);
            });

            Setup.PauseAndAbort(command, assertAfterPause);

            Assert.AreEqual(State.Aborted, command.CurrentState);
            Assert.AreEqual(State.Aborted, command.GetSubCommand(2).CurrentState);
        }

        [TestMethod()]
        public void ExternalAbortTest()
        {
            SimpleCommand p1, p2, p3, p4, p5, p6, p7;
            var command = new ParallelCommand()
                .Add(p1 = new SimpleCommand(() => Thread.Sleep(100), "P1"))
                .Add(p2 = new SimpleCommand(() => Thread.Sleep(100), "P2"))
                .Add(p3 = new SimpleCommand(() => Thread.Sleep(500), "P3"))
                .Add(p4 = new SimpleCommand(() => Thread.Sleep(400), "P4"))
                .Add(p5 = new SimpleCommand(() => Thread.Sleep(600), "P5"))
                .Add(p6 = new SimpleCommand(() => Thread.Sleep(200), "P6"))
                .Add(p7 = new SimpleCommand(() => Thread.Sleep(600), "P7"));

            Setup.RunAndAbort(command);

            // Sub-command are not aborted, they all should be completed by the time abort has taken effect
            Assert.AreEqual(State.Aborted, command.CurrentState);
            Assert.AreEqual(State.Completed, p1.CurrentState);
            Assert.AreEqual(State.Completed, p2.CurrentState);
            Assert.AreEqual(State.Completed, p3.CurrentState);
            Assert.AreEqual(State.Completed, p4.CurrentState);
            Assert.AreEqual(State.Completed, p5.CurrentState);
            Assert.AreEqual(State.Completed, p6.CurrentState);
            Assert.AreEqual(State.Completed, p7.CurrentState);

            Setup.RunAndAbort(command);

            // Sub-command are not aborted, they all should be completed by the time abort has taken effect
            Assert.AreEqual(State.Aborted, command.CurrentState);
            Assert.AreEqual(State.Completed, p1.CurrentState);
            Assert.AreEqual(State.Completed, p2.CurrentState);
            Assert.AreEqual(State.Completed, p3.CurrentState);
            Assert.AreEqual(State.Completed, p4.CurrentState);
            Assert.AreEqual(State.Completed, p5.CurrentState);
            Assert.AreEqual(State.Completed, p6.CurrentState);
            Assert.AreEqual(State.Completed, p7.CurrentState);
        }

        [TestMethod()]
        public void RetrieveSubCommandsTest()
        {
            var command1 = new ParallelCommand("Branch 1")
                .Add(new SimpleCommand("B1-1"))
                .Add(new SimpleCommand("B1-2"))
                .Add(new SimpleCommand("B1-3"));

            var command2 = new ParallelCommand("Branch 2")
                .Add(new SimpleCommand("B2-1"))
                .Add(new SimpleCommand("B2-2"))
                .Add(new SimpleCommand("B2-3"))
                .Add(new SimpleCommand("B2-4"))
                .Add(new SimpleCommand("B2-5"));

            var command = new ParallelCommand("Main")
                .Add(new SimpleCommand("A1"))
                .Add(new SimpleCommand("A2"))
                .Add(command1)
                .Add(command2);

            Assert.AreEqual(3, command1.Descendants.Count());
            Assert.AreEqual(5, command2.Descendants.Count());
            Assert.AreEqual(12, command.Descendants.Count());
            Assert.AreEqual(3, command1.Children.Count());
            Assert.AreEqual(5, command2.Children.Count());
            Assert.AreEqual(4, command.Children.Count());
        }

        [TestMethod()]
        public void RunInputOutputTest()
        {
            var command = new SimpleCommandIO<string, int>(input => input.Length, "Test");
            var parallelCommand = new ParallelCommand();

            parallelCommand.Add(command);

            command.Input = "input";
            parallelCommand.Run();
        }

        //----------------------------------------------------------------------------------------------------------------------

        private ParallelCommand CreatePauseAbortParallelCommand()
        {
            var command = new ParallelCommand("Parallel test command");
            var command1 = new SequentialCommand("P3")
                .Add(new SimpleCommand(() => Thread.Sleep(Setup.ThreadLatencyDelayMsec), "P3-S1"))
                .Add(new SimpleCommand(() => Thread.Sleep(2 * Setup.ThreadLatencyDelayMsec), "P3-S2"))
                .Add(new SimpleCommand(() => Thread.Sleep(4 * Setup.ThreadLatencyDelayMsec), "P3-S3"));

            var command2 = new SequentialCommand("P6");
            command2.Add(new SimpleCommand(() => Thread.Sleep((int)(3 * Setup.ThreadLatencyDelayMsec)), "P6-S1"));
            command2.Add(new SimpleCommand(() => command.Pause(), "P6-S2"));
            command2.Add(new SimpleCommand(() => Thread.Sleep(Setup.ThreadLatencyDelayMsec), "P6-S3"));

            command.Add(new SimpleCommand(() => Thread.Sleep((int)(0.1 * Setup.ThreadLatencyDelayMsec)), "P1"))
                .Add(new SimpleCommand(() => Thread.Sleep((int)(0.1 * Setup.ThreadLatencyDelayMsec)), "P2"))
                .Add(command1)
                .Add(new SimpleCommand(() => Thread.Sleep(7 * Setup.ThreadLatencyDelayMsec), "P4"))
                .Add(new SimpleCommand(() => Thread.Sleep(9 * Setup.ThreadLatencyDelayMsec), "P5"))
                .Add(command2);
            return command;
        }
    }
}

File: Java/ExtensibleCommands/src/test/java/org/extensiblecommands/ParallelCommandTest.java
package org.extensiblecommands;

import org.junit.*;
import org.junit.rules.TestName;

import static java.lang.Thread.sleep;

public class ParallelCommandTest {
    @Rule
    public TestName name = new TestName();

    @BeforeClass
    public static void setUpClass() {
        Setup.InitLog();
    }

    @Before
    public void setUpTest() {
        Logger.log(Logger.LogLevel.Info,
                "----------------------------------------------------------------------------------------------------------");
        Logger.log(Logger.LogLevel.Info,
                String.format("Starting Test %s:%s", this.getClass().getName(), name.getMethodName()));
    }

    @Test
    public void modifyWhileExecutingTest() throws Exception {
        boolean exceptionCaught = false;
        var command = new ParallelCommand("P")
                .add(new AbortableSleepCommand(Setup.ThreadLatencyDelayMsec))
                .add(new AbortableSleepCommand(Setup.ThreadLatencyDelayMsec));

        new Thread(() -> {
            try {
                command.run();
            }
            catch (Exception e) {
                // Ignore
            }
        }).start();
        sleep((int)(0.5*Setup.ThreadLatencyDelayMsec));

        try {
            command.add(SimpleCommand.NullCommand);
        }
        catch (Exception e) {
            if (e.getMessage().contains("Attempt to add"))
                exceptionCaught = true;
        }
        Assert.assertTrue(exceptionCaught);
    }

    @Test
    public void runOkTest() throws Exception {
        var command = new ParallelCommand("Parallel")
            .add(new SimpleCommand(() -> sleep(200), "P1"))
            .add(new SimpleCommand(() -> sleep(300), "P2"))
            .add(new SimpleCommand(() -> sleep(400), "P3"))
            .add(new SimpleCommand(() -> sleep(500), "P4"))
            .add(new SimpleCommand(() -> sleep(600), "P5"))
            .add(new SimpleCommand(() -> sleep(600), "P6"))
            .add(new SimpleCommand(() -> sleep(600), "P7"))
            .add(new SimpleCommand(() -> sleep(600), "P8"))
            .add(new SimpleCommand(() -> sleep(600), "P9"));

        Setup.runAndWaitForNormalCompletion(command);

        Assert.assertTrue(command.getElapsedTimeMsec() < 1500);
    }

    @Test
    public void runErrorTest() throws Exception {
        var command = new ParallelCommand("Parallel")
            .add(new SimpleCommand(() ->sleep(200), "P1"))
            .add(new SimpleCommand(() ->sleep(300), "P2"))
            .add(new SimpleCommand(() -> { throw new ExtensibleCommandsException(Setup.TestErrorCode, Setup.TestErrorDescription); }, "P3-Error"))
            .add(new SimpleCommand(() -> sleep(500), "P4"))
            .add(new SimpleCommand(() -> sleep(600), "P5"))
            .add(new SimpleCommand(() -> sleep(600), "P6"))
            .add(new SimpleCommand(() -> sleep(600), "P7"));

        Setup.runAndWaitForFailure(command);

        Assert.assertTrue(command.getElapsedTimeMsec() < 1500);
        Assert.assertEquals(State.Completed, command.getSubCommand(0).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(1).getState());
        Assert.assertEquals(State.Failed,    command.getSubCommand(2).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(3).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(4).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(5).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(6).getState());

        Assert.assertEquals(Setup.TestErrorCode, command.getException().getId());
        Assert.assertEquals(Setup.TestErrorDescription, command.getException().getText());
    }

    @Test
    public void runFatalErrorTest() throws Exception {
        var command = new ParallelCommand("Parallel")
                .add(new SimpleCommand(() ->sleep(200), "P1"))
                .add(new SimpleCommand(() ->sleep(300), "P2"))
                .add(new SimpleCommand(() -> { throw new Exception(Setup.TestErrorDescription); }, "P3-Error"))
                .add(new SimpleCommand(() -> sleep(500), "P4"))
                .add(new SimpleCommand(() -> sleep(600), "P5"))
                .add(new SimpleCommand(() -> sleep(600), "P6"))
                .add(new SimpleCommand(() -> sleep(600), "P7"));

        try {
            command.run();
        }
        catch (Exception ex) {
            Assert.assertTrue(ex.getMessage().startsWith("Fatal error"));
            Assert.assertEquals(ex.getCause().getMessage(), Setup.TestErrorDescription);
        }

        Setup.runAndWaitForFailure(command);

        Assert.assertTrue(command.getElapsedTimeMsec() < 1500);
        Assert.assertEquals(State.Completed, command.getSubCommand(0).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(1).getState());
        Assert.assertEquals(State.Failed,    command.getSubCommand(2).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(3).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(4).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(5).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(6).getState());
    }

    @Test
    public void abortTest() throws Exception {
        var command = new ParallelCommand("P");
        command.add(new SimpleCommand(() -> sleep(100), "P1"));
        command.add(new SimpleCommand(() -> sleep(100), "P2"));
        command.add(new SimpleCommand(() -> command.abort(), "P3-abort"));
        command.add(new SimpleCommand(() -> sleep(500), "P4"));
        command.add(new SimpleCommand(() -> sleep(400), "P5"));

        Setup.runAndWaitForAbort(command);

        Assert.assertEquals(State.Aborted, command.getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(0).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(1).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(2).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(3).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(4).getState());
    }

    @Test
    public void abortTestWithAbortableSubCommands() throws Exception {

        var command = new ParallelCommand("P");
        command.add(new AbortableSleepCommand(2600, "P1"));
        command.add(new AbortableSleepCommand(2700, "P2"));
        command.add(new SimpleCommand(() ->{ sleep(3*Setup.ThreadLatencyDelayMsec); command.abort();}, "P3-abort"));
        command.add(new AbortableSleepCommand(3700, "P4"));

        Setup.runAndWaitForAbort(command);
        Assert.assertEquals(State.Aborted, command.getState());
        Assert.assertEquals(State.Aborted, command.getSubCommand(0).getState());
        Assert.assertEquals(State.Aborted, command.getSubCommand(1).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(2).getState());
        Assert.assertEquals(State.Aborted, command.getSubCommand(3).getState());
    }

    @Test
    public void pauseResumeTest() throws Exception {
        var command = createPauseAbortParallelCommand();

        ExecutionDelegate assertAfterPause = () -> {
                Assert.assertEquals(State.Executing, command.getState());
                Assert.assertEquals(State.Completed, command.getSubCommand(0).getState());
                Assert.assertEquals(State.Completed, command.getSubCommand(1).getState());
                Assert.assertEquals(State.Executing, command.getSubCommand(2).getState());
                Assert.assertEquals(State.Executing, command.getSubCommand(3).getState());
                Assert.assertEquals(State.Executing, command.getSubCommand(4).getState());
                Assert.assertEquals(State.Executing, command.getSubCommand(5).getState());
            };

        Setup.pauseAndResume(command, assertAfterPause);

        //Assert.assertEquals(State.Completed, command.getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(0).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(1).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(2).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(3).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(4).getState());
        Assert.assertEquals(State.Completed, command.getSubCommand(5).getState());
    }

    @Test
    public void pauseAbortTest() throws Exception {
        var command = createPauseAbortParallelCommand();

        ExecutionDelegate assertAfterPause = () -> {
            Assert.assertEquals(State.Executing, command.getState());
            Assert.assertEquals(State.Completed, command.getSubCommand(0).getState());
            Assert.assertEquals(State.Completed, command.getSubCommand(1).getState());
            Assert.assertEquals(State.Executing, command.getSubCommand(2).getState());
            Assert.assertEquals(State.Executing, command.getSubCommand(3).getState());
            Assert.assertEquals(State.Executing, command.getSubCommand(4).getState());
            Assert.assertEquals(State.Executing, command.getSubCommand(5).getState());
        };

        Setup.pauseAndAbort(command, assertAfterPause);

        Assert.assertEquals(State.Aborted, command.getState());
        Assert.assertEquals(State.Aborted, command.getSubCommand(2).getState());
    }

    @Test
    public void externalAbortTest() throws Exception {
        SimpleCommand p1, p2, p3, p4, p5, p6, p7;
        var command = new ParallelCommand("P")
            .add(p1 = new SimpleCommand(() ->sleep(100), "P1"))
            .add(p2 = new SimpleCommand(() ->sleep(100), "P2"))
            .add(p3 = new SimpleCommand(() ->sleep(500), "P3"))
            .add(p4 = new SimpleCommand(() ->sleep(400), "P4"))
            .add(p5 = new SimpleCommand(() ->sleep(600), "P5"))
            .add(p6 = new SimpleCommand(() ->sleep(200), "P6"))
            .add(p7 = new SimpleCommand(() ->sleep(600), "P7"));

        Setup.runAndAbort(command);

        // Sub-command are not aborted, they all should be completed by the time abort has taken effect
        Assert.assertEquals(State.Aborted, command.getState());
        Assert.assertEquals(State.Completed, p1.getState());
        Assert.assertEquals(State.Completed, p2.getState());
        Assert.assertEquals(State.Completed, p3.getState());
        Assert.assertEquals(State.Completed, p4.getState());
        Assert.assertEquals(State.Completed, p5.getState());
        Assert.assertEquals(State.Completed, p6.getState());
        Assert.assertEquals(State.Completed, p7.getState());

        Setup.runAndAbort(command);

        // Sub-command are not aborted, they all should be completed by the time abort has taken effect
        Assert.assertEquals(State.Aborted, command.getState());
        Assert.assertEquals(State.Completed, p1.getState());
        Assert.assertEquals(State.Completed, p2.getState());
        Assert.assertEquals(State.Completed, p3.getState());
        Assert.assertEquals(State.Completed, p4.getState());
        Assert.assertEquals(State.Completed, p5.getState());
        Assert.assertEquals(State.Completed, p6.getState());
        Assert.assertEquals(State.Completed, p7.getState());
    }

    @Test
    public void retrieveSubCommandsTest() {
        var command = new ParallelCommand("Main")
            .add(new SimpleCommand("A1"))
            .add(new SimpleCommand("A2"));

        var command1 = new ParallelCommand("Branch 1")
            .add(new SimpleCommand("B1-1"))
            .add(new SimpleCommand("B1-2"))
            .add(new SimpleCommand("B1-3"));

        var command2 = new ParallelCommand("Branch 2")
            .add(new SimpleCommand("B2-1"))
            .add(new SimpleCommand("B2-2"))
            .add(new SimpleCommand("B2-3"))
            .add(new SimpleCommand("B2-4"))
            .add(new SimpleCommand("B2-5"));

        command.add(command1);
        command.add(command2);

        Assert.assertEquals(3, command1.getDescendants().spliterator().getExactSizeIfKnown());
        Assert.assertEquals(5, command2.getDescendants().spliterator().getExactSizeIfKnown());
        Assert.assertEquals(12, command.getDescendants().spliterator().getExactSizeIfKnown());
        Assert.assertEquals(3, command1.getChildren().spliterator().getExactSizeIfKnown());
        Assert.assertEquals(5, command2.getChildren().spliterator().getExactSizeIfKnown());
        Assert.assertEquals(4, command.getChildren().spliterator().getExactSizeIfKnown());
    }

    @Test
    public void runInputOutputTest() throws Exception {
        var command = new SimpleCommandIO<String, Integer>(input -> input.length(), "Test");

        var parallelCommand = new ParallelCommand();
        parallelCommand.add(command);

        command.setInput("input");
        parallelCommand.run();
    }

    //----------------------------------------------------------------------------------------------------------------------

    private ParallelCommand createPauseAbortParallelCommand() {
        var command = new ParallelCommand("Parallel test command");
        var command1 = new SequentialCommand("P3")
            .add(new SimpleCommand(() -> sleep(Setup.ThreadLatencyDelayMsec), "P3-S1"))
            .add(new SimpleCommand(() -> sleep(2*Setup.ThreadLatencyDelayMsec), "P3-S2"))
            .add(new SimpleCommand(() -> sleep(4*Setup.ThreadLatencyDelayMsec), "P3-S3"));

        var command2 = new SequentialCommand("P6")
            .add(new SimpleCommand(() -> sleep((int)(3*Setup.ThreadLatencyDelayMsec)), "P6-S1"));
        command2.add(new SimpleCommand(() -> command.pause(), "P6-S2"));
        command2.add(new SimpleCommand(() -> sleep(Setup.ThreadLatencyDelayMsec), "P6-S3"));

        command.add(new SimpleCommand(() -> sleep((int)(0.1 * Setup.ThreadLatencyDelayMsec)), "P1"))
            .add(new SimpleCommand(() -> sleep((int)(0.1 * Setup.ThreadLatencyDelayMsec)), "P2"))
            .add(command1)
            .add(new SimpleCommand(() -> sleep(7 * Setup.ThreadLatencyDelayMsec), "P5"))
            .add(new SimpleCommand(() -> sleep(9*Setup.ThreadLatencyDelayMsec), "P4"))
            .add(command2);
        return command;
    }
}
